### ISSUE TITLE
Problem with GovernmentCooldownChanges()

### ISSUE THREAD

**Issue by JFDaly** [LoneGazebo/Community-Patch-DLL#1166]
GameEvent.GovernmentCooldownChanges(), one added to the DLL specifically for Sovereingty, seems to have a slight problem. The second parameter which should pass the new cooldown (retrievable via player:GetGovernmentCooldown()) appears to pass the Cooldown rate (retrievable via player:GetGovernmentCooldownRate()) instead. This is not an issue for the ReformCooldownChanges game event, which correctly passes the new cooldown value. 

The confusing thing, however, is that the event registers SetGovernmentCooldown correctly - i.e. if I directly invoke player:SetGovernmentCooldown(2), it'll pass 2. It's only when the Cooldown subtracts naturally that it passes the CooldownRate - i.e. each turn, the DLL automatically subtracts from the cooldown according to GetCooldownRate, but that returns, for instance, -1 each time.

Would appreciate it if you could look into this. Thanks.

**Comment by LoneGazebo** [comment]
Looking at it now, and it seems to function correctly. The code is here:

```
void CvPlayer::ChangeGovernmentCooldown(int iValue)
{
	GAMEEVENTINVOKE_HOOK(GAMEEVENT_GovernmentCooldownChanges, GetID(), iValue);
	m_iJFDGovernmentCooldown += iValue;
}
int CvPlayer::GetGovernmentCooldown() const
{
	return m_iJFDGovernmentCooldown;
}
void CvPlayer::SetGovernmentCooldown(int iValue)
{
	GAMEEVENTINVOKE_HOOK(GAMEEVENT_GovernmentCooldownChanges, GetID(), iValue);
	if(m_iJFDGovernmentCooldown != iValue)
	{
		m_iJFDGovernmentCooldown = iValue;
	}
}

void CvPlayer::ChangeGovernmentCooldownRate(int iValue)
{
	GAMEEVENTINVOKE_HOOK(GAMEEVENT_GovernmentCooldownRateChanges, GetID(), iValue);
	m_iJFDGovernmentCooldownRate += iValue;
}
int CvPlayer::GetGovernmentCooldownRate() const
{
	return m_iJFDGovernmentCooldownRate;
}
void CvPlayer::SetGovernmentCooldownRate(int iValue)
{
	GAMEEVENTINVOKE_HOOK(GAMEEVENT_GovernmentCooldownRateChanges, GetID(), iValue);
	if(m_iJFDGovernmentCooldownRate != iValue)
	{
		m_iJFDGovernmentCooldownRate = iValue;
	}
}
```

Not sure...

**Comment by JFDaly** [comment]
Hm, I'll run some more tests then, and keep you posted.

**Comment by LoneGazebo** [comment]
Cool. If my brain is missing something above, let me know.

G

**Comment by JFDaly** [comment]
What's the code that has GovernmentCooldown change each turn in the DLL according to the cooldown rate? I don't see it above, and that's the part that doesn't seem to be working correctly (Change/SetGovernmentCooldown through Lua works fine).

**Comment by LoneGazebo** [comment]
`void CvPlayer::DoGovernmentCooldown()
{
	if(GetGovernmentCooldown() <= 0)
	{
		return;
	}
	int iRate = GetGovernmentCooldownRate();
	if(iRate != 0)
	{
		ChangeGovernmentCooldown(-iRate);
		if(GetGovernmentCooldown() <= 0)
		{
			SetGovernmentCooldown(0);
		}
	}
}`

**Comment by LoneGazebo** [comment]
FYI, use the <IsDummy>true</IsDummy> XML element for your dummy policies - the 'no_branch_type' element is a no-go because that would include opener/finisher policy elements.

G

**Comment by JFDaly** [comment]
Ah, sorry, was an oversight on my part. I'll stick with IsDummy.

**Comment by LoneGazebo** [comment]
Any update on the cooldown issue?

**Comment by JFDaly** [comment]
Still same issue: second parameter, which should be the current cooldown turns, sends the inverse of the Government cooldown rate (i.e. -1 if 1, -2 if 2). At 5 cooldown turns, I send a notification, but this never comes to pass because it is always -1. But when the cooldown would turn to 0, this successfully triggers the main function of this event (the reshuffling of the Legislature) - although it does not allow me to reset the Government Cooldown, which could be because of the time at which the event triggers.

I can check GetGovernmentCooldown() in that event to successfully retrieve the cooldown, but doing so results in an infinite loop of some sorts. I can use DoTurn to get around this, but that would render the event unusable, which would be unfortunate (and means I need to take extra steps in order to reshuffle the Legislature if I want to do so elsewhere).

tl;dr No real update, no.

**Comment by LoneGazebo** [comment]
Ooooh, I think I see. You want the event the push back the new value for the int (i.e. if the event is for ChangeGovernmentCooldown, you want the new cooldown, not the amount it was cooled-down-by. Correct? I can fix that.

**Comment by JFDaly** [comment]
Yeah, that's what I'm looking for; like how ReformCooldown presently works.

**Comment by LoneGazebo** [comment]
Try my latest beta (just posted it), test both reform and government, see if they're working as intended.

G

**Comment by JFDaly** [comment]
'Fraid not. GameEvent.ReformCooldownChanges no longer functions, and GovernmentCooldownChanges returns some kind of ID(?) in place of the cooldown: 1512370252

**Comment by LoneGazebo** [comment]
Oh, must be because LUA m_i values. Of course. Damn. Okay, I'll redo it in a bit (heading out for awhile) and get back to you. Sorry!

**Comment by JFDaly** [comment]
No worries - thanks for looking into it.

**Comment by LoneGazebo** [comment]
[(1) Community Patch (JFD Beta).zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/110656/1.Community.Patch.JFD.Beta.zip)

Try this guy, see if it works.

**Comment by JFDaly** [comment]
The only - minor - issue is that the ReformCooldownChanges event seems to fire twice when the change reaches 0, thus resulting in the function firing twice (in this case, sending two notifications that the cooldown has ended). The same doesn't happen for the Government Cooldown. Otherwise, it's working great!

**Comment by LoneGazebo** [comment]
Hmm...I can set a bool to prevent that, if that'd be helpful.

**Comment by LoneGazebo** [comment]
Glad it is working now.

**Comment by JFDaly** [comment]
If that'd help with the issue, yeah - thanks.
